src/tensor_ops/utilities/reduction_utils.rs

```diff
@@ -1,8 +1,48 @@
-#[cfg(feature = "cuda")]
-use crate::prelude::{Axes, Shape};
+use crate::shapes::{Axes, Shape};
+use crate::tensor::cpu::NdIndex;
 #[cfg(feature = "cuda")]
 use std::vec::Vec;
 
+/// Permutes strides so that all reduces axes are rightmost.
+/// For example reducing (2, 3, 4) to (4,) would permute the shape to (4, 2, 3)
+/// via the strides.
+///
+/// This makes all the reduced elements sequential when indexed using [NdIndex].
+#[inline(always)]
+pub(crate) fn index_for_reductions<S: Shape, Ax: Axes>(
+    shape: S,
+    strides: S::Concrete,
+) -> NdIndex<S> {
+    let dims = shape.concrete();
+    let mut new_shape: S::Concrete = Default::default();
+    let mut new_strides: S::Concrete = Default::default();
+    let num_non_reduced_dims = S::NUM_DIMS - Ax::as_array().into_iter().count();
+
+    let mut i_reduced = 0;
+    let mut i_non_reduced = 0;
+    for i_src in 0..S::NUM_DIMS {
+        if Ax::as_array().into_iter().any(|x| x == i_src as isize) {
+            // this axis is reduced
+            new_shape[num_non_reduced_dims + i_reduced] = dims[i_src];
+            new_strides[num_non_reduced_dims + i_reduced] = strides[i_src];
+            i_reduced += 1;
```

Comment:
An optimization for this is probably to sort the strides in this section of `new_shape`/`new_strides` by decreasing size, to maximize locality when indexing.